src/rust/core/server/src/threads/admin.rs

```diff
@@ -0,0 +1,365 @@
+    // other backends
+    pub fn get_rusage(&self) {
+        let mut rusage = libc::rusage {
+            ru_utime: libc::timeval {
+                tv_sec: 0,
+                tv_usec: 0,
+            },
+            ru_stime: libc::timeval {
+                tv_sec: 0,
+                tv_usec: 0,
+            },
+            ru_maxrss: 0,
+            ru_ixrss: 0,
+            ru_idrss: 0,
+            ru_isrss: 0,
+            ru_minflt: 0,
+            ru_majflt: 0,
+            ru_nswap: 0,
+            ru_inblock: 0,
+            ru_oublock: 0,
+            ru_msgsnd: 0,
+            ru_msgrcv: 0,
+            ru_nsignals: 0,
+            ru_nvcsw: 0,
+            ru_nivcsw: 0,
+        };
+
+        if unsafe { libc::getrusage(libc::RUSAGE_SELF, &mut rusage) } == 0 {
+            set_counter!(
+                &Stat::RuUtime,
+                rusage.ru_utime.tv_sec as u64 * 1000000000 + rusage.ru_utime.tv_usec as u64 * 1000,
+            );
+            set_counter!(
+                &Stat::RuStime,
+                rusage.ru_stime.tv_sec as u64 * 1000000000 + rusage.ru_stime.tv_usec as u64 * 1000,
+            );
+            set_gauge!(&Stat::RuMaxrss, rusage.ru_maxrss as i64);
+            set_gauge!(&Stat::RuIxrss, rusage.ru_ixrss as i64);
+            set_gauge!(&Stat::RuIdrss, rusage.ru_idrss as i64);
+            set_gauge!(&Stat::RuIsrss, rusage.ru_isrss as i64);
+            set_counter!(&Stat::RuMinflt, rusage.ru_minflt as u64);
+            set_counter!(&Stat::RuMajflt, rusage.ru_majflt as u64);
+            set_counter!(&Stat::RuNswap, rusage.ru_nswap as u64);
+            set_counter!(&Stat::RuInblock, rusage.ru_inblock as u64);
+            set_counter!(&Stat::RuOublock, rusage.ru_oublock as u64);
+            set_counter!(&Stat::RuMsgsnd, rusage.ru_msgsnd as u64);
+            set_counter!(&Stat::RuMsgrcv, rusage.ru_msgrcv as u64);
+            set_counter!(&Stat::RuNsignals, rusage.ru_nsignals as u64);
+            set_counter!(&Stat::RuNvcsw, rusage.ru_nvcsw as u64);
+            set_counter!(&Stat::RuNivcsw, rusage.ru_nivcsw as u64);
+        }
+    }
+}
+
+impl EventLoop for Admin {
+    fn handle_data(&mut self, token: Token) -> Result<(), std::io::Error> {
+        trace!("handling request for admin session: {}", token.0);
+        if let Ok(session) = self.poll.get_mut_session(token) {
+            loop {
+                if session.write_capacity() == 0 {
+                    // if the write buffer is over-full, skip processing
+                    break;
+                }
+                match Parse::parse(session.buffer()) {
+                    Ok(parsed_request) => {
+                        let consumed = parsed_request.consumed();
+                        let request = parsed_request.into_inner();
+                        session.consume(consumed);
+                        match request {
+                            AdminRequest::Stats => {
```

Comment:
I moved handling of stats and version requests out into their own functions. This at least improves this particular section. It's possible we can improve further, but when we move to HTTP admin, this will likely all be replaced?

Reply:
This is perfect for now and easy to change later